src/edge/simulator.ts

```diff
@@ -226,10 +194,148 @@ export class Simulator {
         });
     }
 
+    private async getLastestStandaloneSimulatorInfo() {
+        if (!this.latestStandaloneSimulatorInfoJson) {
+            const infoString = await request.get(Simulator.latestReleaseInfoUrl, {
+                headers: {
+                    "User-Agent": "vscode-azure-iot-edge",
+                },
+            });
+            const infoJson = JSON.parse(infoString);
+            this.latestStandaloneSimulatorInfoJson = infoJson;
+            return infoJson;
+        } else {
+            return this.latestStandaloneSimulatorInfoJson;
+        }
+    }
+
+    private async getSimulatorType(): Promise<SimulatorType> {
+        if (await this.simulatorInstalled()) {
+            if (Simulator.currentPlatform === "win32") {
+                return SimulatorType.Standalone;
+            } else {
+                return SimulatorType.Pip;
+            }
+        }
+        return SimulatorType.NotInstalled;
+    }
+
+    private async getLatestSimulatorVersion(): Promise<string | undefined> {
+        try {
+            let version: string;
+            if (Simulator.currentPlatform === "win32") {
+                const releaseInfoJson: any = await this.getLastestStandaloneSimulatorInfo();
+                version = releaseInfoJson.tag_name;
+            } else {
+                const pipResponse: string = await request.get(Simulator.iotedgehubdevVersionUrl);
+                version = JSON.parse(pipResponse).info.version;
+            }
+            return version;
+        } catch (error) {
+            return undefined;
+        }
+    }
+
+    private async getCurrentSimulatorVersion(): Promise<string | undefined> {
+        const output: string = await Executor.executeCMD(undefined, this.getSimulatorExecutorPath(true), { shell: true }, "--version");
+        const version: string | null = Simulator.extractVersion(output);
+        return version;
+    }
+
+    private async simulatorInstalled(): Promise<boolean> {
+        return await Simulator.checkCmdExist(this.getSimulatorExecutorPath(true));
+    }
+
+    private getSimulatorExecutorPath(forceUseCmd: boolean = false): string {
+        let executorPath: string = Simulator.simulatorExcutableName;
+        if (Simulator.currentPlatform === "win32") {
+            if (this.standaloneSimulatorInstalledPath) {
+                if (!forceUseCmd) {
+                    executorPath = `"${Utility.adjustFilePath(this.standaloneSimulatorInstalledPath)}"`;
+                    if (Utility.isUsingPowershell()) {
+                        executorPath = `& ${executorPath}`;
+                    }
+                } else {
+                    executorPath = `"${this.standaloneSimulatorInstalledPath}"`;
+                }
+            } else {
+                return "";
```

Comment:
return null or undefined to be more clear.

Reply:
OK